### PR TITLE
Port Naming styles analyzer/fixer/tests to shared layer

### PR DESCRIPTION
**Pull request by mavasani**
Strongly recommend reviewing the PR commit by commit - I have tried to add descriptive commit messages to help in the review and will also leave some comments on the PR wherever relevant.

**NOTE:** #42218 blocks enabling the ported code fixer in the CodeStyle layer, but the fixer has been ported without the Export attribute to enable running the unit tests.

### CODE REVIEW

File: src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,3 +1,33 @@
+Microsoft.CodeAnalysis.Internal.Editing.DeclarationModifiers
```
Comment (mavasani) [inline src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt]: This is a public type defined in Workspaces, which is also needed for NamingStyle rules. It has been linked into the CodeStyle analyzers project under a different namespace `Microsoft.CodeAnalysis.Internal.Editing` versus `Microsoft.CodeAnalysis.Editing`

File: src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    internal static class NamingStyleOptions
+    {
+        // Use 'SimplificationOptions' for back compat as the below option 'NamingPreferences' was defined with feature name 'SimplificationOptions'.
+        private const string FeatureName = "SimplificationOptions";
+
+        /// <summary>
+        /// This option describes the naming rules that should be applied to specified categories of symbols, 
+        /// and the level to which those rules should be enforced.
+        /// </summary>
+        internal static PerLanguageOption<NamingStylePreferences> NamingPreferences { get; } = new PerLanguageOption<NamingStylePreferences>(FeatureName, nameof(NamingPreferences), defaultValue: NamingStylePreferences.Default,
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs]: This internal option was just moved from a public Workspaces type `SimplificationOptions` into a separate type `NamingStyleOptions` so this file can be used in the CodeStyle layer.

File: src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs [hunk 3]
```diff
@@ -15,24 +15,6 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics.NamingStyle
 {
     public class EditorConfigNamingStyleParserTests
     {
-        internal static SymbolKindOrTypeKind ToSymbolKindOrTypeKind(object symbolOrTypeKind)
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/Diagnostics/NamingStyles/EditorConfigNamingStyleParserTests.cs]: Just moved to a shared test file.

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs [hunk 4]
```diff
@@ -205,6 +207,13 @@ static string GenerateAnalyzerConfigText(IDictionary<OptionKey, object> options)
 
                             Assert.False(true, "Unexpected non-editorconfig option");
                         }
+                        else if (value is NamingStylePreferences namingStylePreferences)
```
Comment (mavasani) [inline src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs]: This else if part is the core functionality to enable serializing naming style options into an editorconfig file, which is needed for tests running in CodeStyle layer that do not have access to Workspace options.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs [hunk 5]
```diff
@@ -333,6 +333,9 @@ private static string ToEditorConfigString(this SymbolKindOrTypeKind symbol)
                 case TypeKind.Delegate:
                     return "delegate";
 
+                case TypeKind.TypeParameter:
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs]: The change in this file and the following file was a bug that was found in the subsequent commit when tests were ported to code style layer. Attempting to serialize all possible combinations of naming styles to an editorconfig file (which is now done for the tests from code style layer) hits these cases.

File: src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt]: does 'internal' mean we are not signed up to supporting this APi and can remove/break it at any time?

File: src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#else
+using Microsoft.CodeAnalysis.Options;
+namespace Microsoft.CodeAnalysis.Simplification
+#endif
+{
+#if CODE_STYLE
+    public static class NamingStyleOptions
+#else
+    internal static class NamingStyleOptions
+#endif
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs]: this is just reall nasty to read.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs]: My strong preference in a case like this is to have this broken out into its own small PR that can then go through independently.  This PR canthen branchoff of those.

File: src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (mavasani) [reply src/CodeStyle/Core/Analyzers/PublicAPI.Unshipped.txt]: CodeStyle packages are not really libraries, so there is no support. Sam added these files just to track dependencies from this layer.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs (hunk 5 above)
Comment (mavasani) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs]: That was my original intent, but unfortunately the naming tests need to be in CodeStyle layer to fail due to this change, and that is done in the last commit of this PR.

File: src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs (hunk 6 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs]: I know, I’ll see if I can work on the OptionDefinition based design that Sam was proposing to get rid of this Options related mess.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs (hunk 5 above)
Comment (mavasani) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/EditorConfig/EditorConfigNamingStyleParser_SymbolSpec.cs]: @CyrusNajmabadi I was actually able to find a simpler unit test that uncovers these failures. So, I have extracted this bug fix into https://github.com/dotnet/roslyn/pull/42245. Thanks!

**Comment by mavasani** [comment]
@CyrusNajmabadi any other feedback on this one? I'd like to merge this one before working on the next port to avoid merge conflicts. Thanks!

File: src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        private const string FeatureName = "SimplificationOptions";
+
+        /// <summary>
+        /// This option describes the naming rules that should be applied to specified categories of symbols, 
+        /// and the level to which those rules should be enforced.
+        /// </summary>
+        internal static PerLanguageOption<NamingStylePreferences> NamingPreferences { get; } = new PerLanguageOption<NamingStylePreferences>(FeatureName, nameof(NamingPreferences), defaultValue: NamingStylePreferences.Default,
+            storageLocations: new OptionStorageLocation[] {
+                new NamingStylePreferenceEditorConfigStorageLocation(),
+                new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.NamingPreferences5"),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Simplification/NamingStyleOptions.cs]: NamingPreferences5?

File: src/Workspaces/Core/Portable/Simplification/SimplificationOptions.cs [hunk 8]
```diff
@@ -105,16 +104,5 @@ public static class SimplificationOptions
...
         public static PerLanguageOption<bool> PreferIntrinsicPredefinedTypeKeywordInMemberAccess { get; } = new PerLanguageOption<bool>(nameof(SimplificationOptions), nameof(PreferIntrinsicPredefinedTypeKeywordInMemberAccess), defaultValue: true);
-
-        /// <summary>
-        /// This option describes the naming rules that should be applied to specified categories of symbols, 
-        /// and the level to which those rules should be enforced.
-        /// </summary>
-        internal static PerLanguageOption<NamingStylePreferences> NamingPreferences { get; } = new PerLanguageOption<NamingStylePreferences>(nameof(SimplificationOptions), nameof(NamingPreferences), defaultValue: NamingStylePreferences.Default,
-            storageLocations: new OptionStorageLocation[] {
-                new NamingStylePreferenceEditorConfigStorageLocation(),
-                new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.NamingPreferences5"),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Simplification/SimplificationOptions.cs]: hunh... i guess that's here too.

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**